[Docs][Recipes]Sourcing from Wordpress (#17374)
gatsbyjs/gatsby: docs/docs/recipes.md

- [Guide to creating pages from data programmatically](/docs/programmatically-create-pages-from-data/)
- [Example repo](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipe-sourcing-markdown) for this recipe

### Sourcing from Wordpress

#### Prerequisites

- An existing [Gatsby site](/docs/quick-start/) with a `gatsby-config.js` and `gatsby-node.js` file

> - An existing template component such as `src/templates/post.js`

Reviewer: Should be also show the content or link to a file to see what is in the `templates/post.js`?

Author: That does sound helpful! We could add a link to the full-length Reference guide or a blog post as a note. They can refer to it for the template content if they wish.  
This is so that the scope of the recipe is kept brief.
Thoughts?

Reviewer: @gatsbyjs/docs and @marcysutton knows a best practice for the receipts

Author: Awaiting the review from her then

Reviewer 2: I think the best way forward is to create an example for the recipe. Here's some precedent for a Markdown version: https://github.com/gatsbyjs/gatsby/tree/master/examples/recipe-sourcing-markdown

With some code in place that is made intentionally for this recipe, we can link to the `post.js` file in that example in the prerequisites AND give people something to play around with. The [`using-wordpress`](https://github.com/gatsbyjs/gatsby/tree/master/examples/using-wordpress) example is a bit more complicated as it uses Advanced Custom Fields, otherwise I would have recommended that.

Author: That would be great! I have opened a PR for an example [here](/gatsbyjs/gatsby/pull/17425). We can link this once it is approved to give the people more clarity.

#### Directions

1. Install the `gatsby-source-wordpress` plugin by running the following command:

```shell
npm install gatsby-source-wordpress
```

> 2. Configure the plugin by modifying the `gatsby-config.js` file such that it looks like the following:

Reviewer 2: ```suggestion
2. Configure the plugin by modifying the `gatsby-config.js` file such that it includes the following:
```

```JS:title=gatsby-config.js
module.exports = {
  ...
  plugins: [

>     ...,

Reviewer 2: ```suggestion
```
I think removing this is a good idea, so people don't get confused by invalid code!

Reviewer: I like this notation so people don't expect that it is the only plugin to install. In some cases helper plugins are needed

Reviewer: If we make it as a comment then it is valid code?

Reviewer: but I think this discussion is not in the scope of this PR

Author: Makes sense! Thanks!

    {
      resolve: `gatsby-source-wordpress`,
      options: {

>         // your wordpress source

Reviewer 2: ```suggestion
        // baseUrl will need to be updated with your wordpress source
```

        baseUrl: `wpexample.com`,
        protocol: `https`,
        // is it hosted on wordpress.com, or self-hosted?
        hostingWPCOM: false,
        // does your site use the Advanced Custom Fields Plugin?
        useACF: false
      }
    },
  ]
}
```

> **Note**

Reviewer 2: ```suggestion
```
The pattern we've been following is to put the asterisks and note on the same line as the note itself, with a blockquote to italicize. See next line:

Author: Sure! Will keep this new pattern in mind for the future.

> Refer to the [plugin docs](/packages/gatsby-source-wordpress/?=wordpre#how-to-use) to know more about configuring your plugins.

Reviewer: And more hint where the link goes:

```suggestion
Refer to the [`gatsby-source-wordpress` plugin docs](/packages/gatsby-source-wordpress/?=wordpre#how-to-use) to know more about configuring your plugins.
```

> Refer to the [`gatsby-source-wordpress` plugin docs](/packages/gatsby-source-wordpress/?=wordpre#how-to-use) to know more about configuring your plugins.

Reviewer 2: ```suggestion
> > **Note:** Refer to the [`gatsby-source-wordpress` plugin docs](/packages/gatsby-source-wordpress/?=wordpre#how-to-use) to know more about configuring your plugins.
```

> 3. Create a static page for your Wordpress posts by pasting the following sample code in `gatsby-node.js` file:

Reviewer 2: ```suggestion
3. Create dynamic pages for your Wordpress posts by pasting the following sample code in `gatsby-node.js`:
```

```JS:title=gatsby-node.js
const path = require(`path`)
const slash = require(`slash`)

exports.createPages = async ({ graphql, actions }) => {
  const { createPage } = actions

  // query content for WordPress posts
  const result = await graphql(`
    query {
      allWordpressPost {
        edges {
          node {
            id
            slug
          }
        }
      }
    }
  `)

  const postTemplate = path.resolve(`./src/templates/post.js`)
  result.data.allWordpressPost.edges.forEach(edge => {
    createPage({

>       // will be the url for the page

Reviewer 2: ```suggestion
      // `path` will be the url for the page
```

      path: edge.node.slug,
      // specify the component template of your choice
      component: slash(postTemplate),
      // In the ^template's GraphQL query, 'id' will be available
      // as a GraphQL variable to query for this posts's data.
      context: {
        id: edge.node.id,
      },
    })
  })
}
```

> 4. Open the `GraphiQL IDE` at at `localhost:8000/__graphql` and open the Docs or Explorer to observe the queryable fields for `allWordpressPosts`.

Reviewer 2: ```suggestion
4. Open the `GraphiQL IDE` at `localhost:8000/__graphql` and open the Docs or Explorer to observe the queryable fields for `allWordpressPosts`.
```

5. Run `gatsby-develop` to see the newly generated pages and navigate through them.

> **Note**

Reviewer 2: ```suggestion
```

> The static page created above has a unique path for navigating to a particular post, a template component for the posts, and a sample GraphQL query to query your Wordpress content.

Reviewer 2: ```suggestion
> **Note:** The dynamic pages created above have unique paths for navigating to particular posts, a template component for the posts, and a sample GraphQL query to source your Wordpress post content.
```

#### Aditional resources

> - [Getting Started with Wordpress and Gatsby](https://www.gatsbyjs.org/blog/2019-04-26-how-to-build-a-blog-with-wordpress-and-gatsby-part-1/)

Reviewer 2: ```suggestion
- [Getting Started with Wordpress and Gatsby](/blog/2019-04-26-how-to-build-a-blog-with-wordpress-and-gatsby-part-1/)
```

> - More on [Sourcing from Wordpress](https://www.gatsbyjs.org/docs/sourcing-from-wordpress/)

Reviewer: ```suggestion
- More on [Sourcing from Wordpress](/docs/sourcing-from-wordpress/)
```

- [Guide to creating pages from data programmatically](/docs/programmatically-create-pages-from-data/)
- [Example repo](https://github.com/gatsbyjs/gatsby/tree/master/examples/recipe-sourcing-markdown) for this recipe

### Sourcing from WordPress

#### Prerequisites

- An existing [Gatsby site](/docs/quick-start/) with a `gatsby-config.js` and `gatsby-node.js` file

> 

Reviewer 2: ```suggestion
- A WordPress instance, either self-hosted or on Wordpress.com
```
This seems like an important detail!

  query($id: String!) {
    wordpressPost(id: { eq: $id }) {
      title
      content
    site {
      siteMetadata {
        title
        subtitle
      }
    }
  }
`
```

4. Create dynamic pages for your Wordpress posts by pasting the following sample code in `gatsby-node.js`:

```JS:title=gatsby-node.js
const path = require(`path`)
const slash = require(`slash`)

exports.createPages = async ({ graphql, actions }) => {
  const { createPage } = actions

  // query content for WordPress posts
  const result = await graphql(`
    query {
      allWordpressPost {
        edges {
          node {
            id
            slug
          }
        }
      }
    }
  `)

  const postTemplate = path.resolve(`./src/templates/post.js`)
  result.data.allWordpressPost.edges.forEach(edge => {
    createPage({
      // `path` will be the url for the page
      path: edge.node.slug,
      // specify the component template of your choice
      component: slash(postTemplate),
      // In the ^template's GraphQL query, 'id' will be available
      // as a GraphQL variable to query for this posts's data.
      context: {
        id: edge.node.id,
      },
    })
  })
}
```

> 5. Run `gatsby-develop` to see the newly generated pages and navigate through them. 

Author: That would be clearer, I agree! Although wouldn't it be better to keep it that way for consistency, since every step on the Recipes page that mentions running the server uses the same pattern? Let me know what you think.

> 5. Run `gatsby-develop` to see the newly generated pages and navigate through them. 

Reviewer: A) That's why I suggest to change all other commands on this page also. 

B) Or we postpone this change and do an "change it all on receipts" PR just for this change later.

> 5. Run `gatsby-develop` to see the newly generated pages and navigate through them. 

Reviewer: You are right. (B) would be the better way. Every PR should change only one thing. 

So go on with current style and ignore my comment

> 5. Run `gatsby-develop` to see the newly generated pages and navigate through them. 

Author: Changing all the command patterns on the page would be a helpful addition that would lend more clarity to the viewers. However, we can open another PR just for that purpose as you suggested. Since this PR only addresses a specific issue, resolving the former through another PR would be easier to track and review. 
Thanks for pointing it out! 😄

> 5. Run `gatsby-develop` to see the newly generated pages and navigate through them. 

Reviewer: Open a nee issue for this: #1726

> 5. Run `gatsby-develop` to see the newly generated pages and navigate through them. 

Reviewer 2: Agreed that we should keep bigger decisions on formatting apart from this PR, which is focused on a WordPress recipe. @muescha I think that's the wrong link...?

> 5. Run `gatsby-develop` to see the newly generated pages and navigate through them. 

Reviewer: #17626

3. Create dynamic pages for your Wordpress posts by pasting the following sample code in `gatsby-node.js`:

```JS:title=gatsby-node.js
const path = require(`path`)
const slash = require(`slash`)

exports.createPages = async ({ graphql, actions }) => {
  const { createPage } = actions

  // query content for WordPress posts
  const result = await graphql(`
    query {
      allWordpressPost {
        edges {
          node {
            id
            slug
          }
        }
      }
    }
  `)

  const postTemplate = path.resolve(`./src/templates/post.js`)
  result.data.allWordpressPost.edges.forEach(edge => {
    createPage({
      // `path` will be the url for the page
      path: edge.node.slug,
      // specify the component template of your choice
      component: slash(postTemplate),
      // In the ^template's GraphQL query, 'id' will be available
      // as a GraphQL variable to query for this posts's data.
      context: {
        id: edge.node.id,
      },
    })
  })
}
```

4. Open the `GraphiQL IDE` at `localhost:8000/__graphql` and open the Docs or Explorer to observe the queryable fields for `allWordpressPosts`.

> 5. Run `gatsby-develop` to see the newly generated pages and navigate through them.

Author: Totally missed that! Thanks for pointing it out!

> 5. Run `gatsby-develop` to see the newly generated pages and navigate through them.

Author: Also, I agree with you and @muescha on including a look into the post.js template in the recipe itself. Information overload is real and too many links can be both distracting and overwhelming. 
I am updating the directions to include the code in post.js from the new example in place. This will avoid viewers from having to navigate a lot. However, we can still have the links to examples if they need more clarity. 
Thanks for this!

> > **Note:** The dynamic pages created above have unique paths for navigating to particular posts, a template component for the posts, and a sample GraphQL query to source your Wordpress post content.

Author: That actually makes more sense than adding it as a comment. Thanks!

#### Aditional resources

- [Getting Started with Wordpress and Gatsby](/blog/2019-04-26-how-to-build-a-blog-with-wordpress-and-gatsby-part-1/)
- More on [Sourcing from Wordpress](/docs/sourcing-from-wordpress/)

> - [Live example on Sourcing from Wordpress](https://watch-learn.com/series/gatsbyjs-wordpress)

Reviewer 2: The watch-learn example is already linked from the Sourcing docs page. Rather than linking to it again here, we should include a link to your new example in the repo. 
```suggestion
- [Live example on Sourcing from Wordpress](/examples/gatsby-sourcing-wordpress/)
```

Author: Ahh, yes! I will replace it by linking my example once it's approved and merged in the repo.

```shell
npm install gatsby-source-wordpress --save
```

2. Configure the plugin by modifying the `gatsby-config.js` file such that it includes the following:

```JS:title=gatsby-config.js
module.exports = {
  ...
  plugins: [
    {
      resolve: `gatsby-source-wordpress`,
      options: {
        // baseUrl will need to be updated with your wordpress source
        baseUrl: `wpexample.com`,
        protocol: `https`,
        // is it hosted on wordpress.com, or self-hosted?
        hostingWPCOM: false,
        // does your site use the Advanced Custom Fields Plugin?
        useACF: false
      }
    },
  ]
}
```

> **Note:** Refer to the [`gatsby-source-wordpress` plugin docs](/packages/gatsby-source-wordpress/?=wordpre#how-to-use) to know more about configuring your plugins.

3. Create a template component such as `src/templates/post.js` with the following code in it:

```JS:title=post.js
import React, { Component } from "react"
import { graphql } from "gatsby"
import PropTypes from "prop-types"

class Post extends Component {
  render() {
    const post = this.props.data.wordpressPost

    return (
      <>
        <h1>{post.title}</h1>

>         <div>{post.content}</div>

Reviewer 3: ```suggestion
        <div dangerouslySetInnerHTML={{ __html: post.content }} />
```

Since the data returned from WordPress post content is a big chunk of html, displaying it like this makes it easier to read when rendered.

Author: Yes, that will look cleaner! Thanks for the suggestion, @gillkyle!

      </>
    )
  }
}

Post.propTypes = {
  data: PropTypes.object.isRequired,
  edges: PropTypes.array,
}

export default Post

export const pageQuery = graphql`
  query($id: String!) {
    wordpressPost(id: { eq: $id }) {
      title
      content

>     site {

Reviewer 3: Since this section of the query isn't used (site/siteMetadata/title/subtitle) I think it could be used. It would also break in sites that this is attempted on that don't have a title and subtitle defined in that shape in a `gatsby-config`
